File: R/reshape_compensation.R
## Reshape compensation matrix for manual input into FACSDiva

wd = dirname(this.path::here())  # wd = '~/github/R/helperFlowCell'
library('optparse')
suppressPackageStartupMessages(library('logr'))
import::from(tidyr, 'pivot_longer', 'pivot_wider')
import::from(XML, 'saveXML')
import::from(matlib, 'inv')

import::from(file.path(wd, 'R', 'functions', 'preprocessing.R'),
    'spillover_to_xml', .character_only=TRUE)
import::from(file.path(wd, 'R', 'tools', 'file_io.R'),
    'join_many_csv', .character_only=TRUE)
import::from(file.path(wd, 'R', 'tools', 'df_tools.R'),
    'rename_columns', 'reset_index', .character_only=TRUE)
import::from(file.path(wd, 'R', 'tools', 'list_tools.R'),
    'items_in_a_not_b', 'list2matrix', 'move_list_items_to_front', .character_only=TRUE)


# ----------------------------------------------------------------------
# Pre-script settings

# args
option_list = list(
    make_option(c("-i", "--input"), default='data/compensation/Acquisition-defined.csv',
                metavar='data/compensation/Acquisition-defined.csv', type="character",
                help="specify the directory of json files"),

    make_option(c("-o", "--output-dir"), default="data/compensation",
                metavar="data/compensation", type="character",
                help="set the output directory for the data"),

    make_option(c("-s", "--sort"), default='ref/fluorochrome_order.txt',
                metavar='ref/fluorochrome_order.csv', type="character",
                help="csv file containing a list"),

    make_option(c("-t", "--troubleshooting"), default=FALSE, action="store_true",
                metavar="FALSE", type="logical",
                help="enable if troubleshooting to prevent overwriting your files")
)
opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)
troubleshooting = opt[['troubleshooting']]

# Start Log
start_time = Sys.time()
log <- log_open(paste0("reshape_compensation-",
    strftime(start_time, format="%Y%m%d_%H%M%S"), '.log'))
log_print(paste('Script started at:', start_time))

# create output dir
if (!dir.exists(file.path(opt[['output-dir']]))) {
    dir.create(file.path(opt[['output-dir']]), recursive=TRUE)
}


# ----------------------------------------------------------------------
# Preprocessing

spillover_matrix <- read.csv(
    file.path(wd, opt[['input']]),
    row.names=1, header=TRUE, check.names=FALSE
)
channels <- row.names(spillover_matrix)

# sort
if (file.exists(file.path(wd, opt[['sort']]))) {
    matrix_order <- read.csv(file.path(wd, opt[['sort']]), header=FALSE)[['V1']]
    matrix_order <- move_list_items_to_front(matrix_order, colnames(spillover_matrix))
    spillover_matrix <- spillover_matrix[matrix_order, matrix_order]
}

# pivot
withCallingHandlers({
    spillover_table <- as.data.frame(pivot_longer(
        reset_index(spillover_matrix, index_name="- % Fluorochrome"),
        cols=channels,
        names_to = "Fluorochrome",
        values_to = "Spectral Overlap"
    ))
}, warning = function(w) {
    if ( any(grepl("Using an external vector in selections", w)) ) {
        invokeRestart("muffleWarning")
    }
})


# ----------------------------------------------------------------------
# Export Spillover Table

# save
if (!troubleshooting) {

    # cleanup
    diva_spillover <- spillover_table[
        (spillover_table[["Fluorochrome"]] != spillover_table[["- % Fluorochrome"]]),
        c("Fluorochrome", "- % Fluorochrome", "Spectral Overlap")
    ]
    diva_spillover[["Spectral Overlap"]] <- diva_spillover[["Spectral Overlap"]]*100
    diva_spillover[["Spectral Overlap"]] <- round(diva_spillover[["Spectral Overlap"]], 2)

    write.table(
        diva_spillover,
        file = file.path(wd, opt[['output-dir']], 'spillover_table.csv'),
        row.names = FALSE, sep=',',
        na=""
    )
}

# ----------------------------------------------------------------------
# Compensation Matrix
# "The compensation matrix is the inverse of the spillover matrix." - Data File Standard for FCS3.1
# Use this to import compensation matrix directly into FACSDiva
# Experiment > [Import Cytometer Settings] 


compensation_matrix <- inv(as.matrix(spillover_matrix))

# save
if (!troubleshooting) {
    write.table(
        t(as.vector(compensation_matrix)),  # flattened
        file = file.path(wd, opt[['output-dir']], 'compensation_matrix.txt'),
        row.names=FALSE, col.names=FALSE, sep=','
    )
}


# ----------------------------------------------------------------------
# XML file
# Use this to copy directly into Flowjo workspace files

filename <- basename(tools::file_path_sans_ext(opt[['input']]))
spillover_xml <- spillover_to_xml(spillover_table, channels, name=filename)

# save
if (!troubleshooting) {
    invisible(saveXML(spillover_xml,
        prefix='<?xml version="1.0" encoding="UTF-8"?>\n',
        file = file.path(wd, opt[['output-dir']], paste0(filename, '.mtx'))
    ))  # cat() to view
}


end_time = Sys.time()
log_print(paste('Script ended at:', Sys.time()))
log_print(paste("Script completed in:", difftime(end_time, start_time)))
log_close()

File: python/reshape_compensation.py
#!/usr/bin/env python3
"""Reshape compensation matrix for FACSDiva import."""

import os
import sys
import argparse
import logging
from datetime import datetime

import pandas as pd
import numpy as np

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from functions.preprocessing import spillover_to_xml
from tools.df_tools import rename_columns, reset_index
from tools.list_tools import move_list_items_to_front


def main():
    parser = argparse.ArgumentParser(description='Reshape compensation matrix')
    parser.add_argument('-i', '--input', default='data/compensation/Acquisition-defined.csv',
                        help='Input spillover matrix CSV')
    parser.add_argument('-o', '--output-dir', default='data/compensation',
                        help='Output directory')
    parser.add_argument('-s', '--sort', default='ref/fluorochrome_order.txt',
                        help='Fluorochrome order file')
    parser.add_argument('-t', '--troubleshooting', action='store_true')

    args = parser.parse_args()

    # Setup
    wd = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

    # Logging
    start_time = datetime.now()
    logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(message)s')
    logging.info(f'Script started at: {start_time}')

    # Create output directory
    os.makedirs(os.path.join(wd, args.output_dir), exist_ok=True)

    # Read spillover matrix
    spillover_matrix = pd.read_csv(
        os.path.join(wd, args.input),
        index_col=0
    )
    channels = spillover_matrix.index.tolist()

    # Sort if order file exists
    order_file = os.path.join(wd, args.sort)
    if os.path.exists(order_file):
        logging.info('Fluorochrome order file found')
        with open(order_file) as f:
            fluorochrome_order = [line.strip() for line in f]

        fluorochrome_order = [f for f in fluorochrome_order
                              if f in spillover_matrix.columns]
        fluorochrome_order = move_list_items_to_front(
            spillover_matrix.columns.tolist(), fluorochrome_order
        )
        spillover_matrix = spillover_matrix.loc[fluorochrome_order, fluorochrome_order]
        channels = spillover_matrix.index.tolist()

    # Pivot to long format
    spillover_matrix_reset = spillover_matrix.reset_index()
    spillover_matrix_reset.columns = ['- % Fluorochrome'] + channels

    spillover_table = spillover_matrix_reset.melt(
        id_vars='- % Fluorochrome',
        value_vars=channels,
        var_name='Fluorochrome',
        value_name='Spectral Overlap'
    )

    # Export spillover table
    if not args.troubleshooting:
        diva_spillover = spillover_table[
            spillover_table['Fluorochrome'] != spillover_table['- % Fluorochrome']
        ].copy()
        diva_spillover['Spectral Overlap'] = (
            diva_spillover['Spectral Overlap'] * 100
        ).round(2)

        filepath = os.path.join(wd, args.output_dir, 'spillover_table.csv')
        diva_spillover.to_csv(filepath, index=False)

    # Compute compensation matrix (inverse of spillover)
    compensation_matrix = np.linalg.inv(spillover_matrix.values)

    if not args.troubleshooting:
        # Flatten and save
        filepath = os.path.join(wd, args.output_dir, 'compensation_matrix.txt')
        with open(filepath, 'w') as f:
            f.write(','.join(str(round(v, 8)) for v in compensation_matrix.flatten()))

    # Export XML
    filename = os.path.splitext(os.path.basename(args.input))[0]
    xml_content = spillover_to_xml(spillover_table, channels, name=filename)

    if not args.troubleshooting:
        filepath = os.path.join(wd, args.output_dir, f'{filename}.mtx')
        with open(filepath, 'w') as f:
            f.write(xml_content)

    end_time = datetime.now()
    logging.info(f'Script ended at: {end_time}')
    logging.info(f'Script completed in: {end_time - start_time}')


if __name__ == '__main__':
    main()
